sw/otbn/crypto/p384_isoncurve.s

```diff
@@ -0,0 +1,184 @@
+ * @param[in]  dmem[24]: dptr_y, pointer to dmem location containing affine
+ *                               y-coordinate of input point
+ *
+ * clobbered registers: x2, x3, w0 to w5, w10 to w17
+ * clobbered flag groups: FG0
+ */
+ .globl p384_isoncurve
+p384_isoncurve:
+
+  /* setup all-zero reg */
+  bn.xor    w31, w31, w31
+
+  /* load affine x-coordinate of curve point from dmem
+     [w1, w0] <= dmem[dptr_x] = dmem[20] */
+  la        x3, dptr_x
+  lw        x3, 0(x3)
+  li        x2, 0
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2++, 32(x3)
+
+  /* load affine y-coordinate of curve point from dmem
+     [w3, w2] <= dmem[dptr_y] = dmem[24] */
+  la        x3, dptr_y
+  lw        x3, 0(x3)
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2, 32(x3)
+
+  /* load domain parameter p (modulus) from dmem
+     [w13, w12] = p = dmem[p384_p] */
+  li        x2, 12
+  la        x3, p384_p
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2++, 32(x3)
+
+  /* load domain parameter b from dmem
+     [w4, w5] = b = dmem[p384_b] */
+  li        x2, 4
+  la        x3, p384_b
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2++, 32(x3)
+
+  /* y^2 = [w17,w16] <= y*y = [w3,w2]*w[w3,w2] */
+  bn.mov    w10, w2
+  bn.mov    w11, w3
+  bn.mov    w16, w2
+  bn.mov    w17, w3
+  jal       x1, p384_mulmod_p
+
+  /* store result (left side): dmem[dptr_s] <= y^2 = [w17,w16] */
+  la        x3, dptr_s
+  lw        x3, 0(x3)
+  li        x2, 16
+  bn.sid    x2++, 0(x3)
+  bn.sid    x2++, 32(x3)
+
+  /*  x^3 = [w17,w16] <= (x*x)*x = ([w1,w0]*(w1,w0])*[w1,w0] */
+  bn.mov    w10, w0
+  bn.mov    w11, w1
+  bn.mov    w16, w0
+  bn.mov    w17, w1
+  jal       x1, p384_mulmod_p
+  bn.mov    w10, w0
+  bn.mov    w11, w1
+  jal       x1, p384_mulmod_p
+
+  /* for curve P-384, 'a' can be written as a = -3, therefore we subtract
+     x three times from x^3.
+     x^3 + ax  mod p = [w17,w16] <= x^3 -3 x mod p
+                     = [w17,w16] - [w1,w0] - [w1,w0] - [w1,w0] mod [w13,w12] */
+  loopi     3, 6
+    bn.sub    w16, w16, w0
+    bn.subb   w17, w17, w1
+    bn.add    w10, w16, w12
+    bn.addc   w11, w17, w13
+    bn.sel    w16, w10, w16, C
+    bn.sel    w17, w11, w17, C
+
+  /* add domain parameter b
+     x^3 + ax + b mod p = [w17,w16] <= [w17,w16] + [w5,w4] mod [w13,w12] */
+  bn.add    w16, w16, w4
+  bn.addc   w17, w17, w5
+  bn.sub    w10, w16, w12
+  bn.subb   w11, w17, w13
+  bn.sel    w16, w16, w10, C
+  bn.sel    w17, w17, w11, C
+
+  /* store result (right side)
+     dmem[dptr_r] <= x^3 + ax + b mod p = [w17,w16] */
+  la        x3, dptr_r
+  lw        x3, 0(x3)
+  li        x2, 16
+  bn.sid    x2++, 0(x3)
+  bn.sid    x2++, 32(x3)
+
+  ret
+
+.data
+
+/* Right side of Weierstrass equation */
+.globl r
+.balign 32
+r:
+  .zero 64
+
+/* Left side of Weierstrass equation */
+.globl s
+.balign 32
+s:
+  .zero 64
+
+/* Curve point x-coordinate. */
+.globl x
+.weak x
+.balign 32
+x:
+  .zero 64
+
+/* Curve point y-coordinate. */
+.globl y
+.weak y
+.balign 32
+y:
+  .zero 64
+
```

Comment:
nit: since you only reference the pointers, I think you shouldn't need the data buffers here.

Reply:
Never mind, I see below that some parts of the code rely on fixed offsets for these buffers...we should really update how the P384 code handles memory! But we talked about that on slack already, it makes sense to do a separate pass for that :slightly_smiling_face: